packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -377,7 +377,14 @@ function ChildReconciler(shouldTrackSideEffects) {
     element: ReactElement,
     expirationTime: ExpirationTime,
   ): Fiber {
-    if (current !== null && current.type === element.type) {
+    const elementType = element.type;
+    if (
+      current !== null &&
+      (current.type === elementType ||
+        (elementType !== null &&
+          elementType !== undefined &&
+          current.type === elementType._reactResult))
```

Comment:
This is super hot. This means that when we reconcile in a long list, we have to do several comparisons and a missing property lookup (which can't be optimized in this function since it's a new hidden class for each one, and traverses the prototype chain all the way up to Object.prototype).

We also don't typically ensure that wrappers get unwrapped to reconcile their inner value so I don't think that we need to semantically support the case where you switch from an instance of the component to the promise of that same component.

I'd say that its worth the complexity elsewhere to allow for `fiber.type` to be the thennable.